### DISCUSSION TITLE
Should `vertico-multiform` set variables after setup?

### DISCUSSION THREAD

**Discussion by aikrahguzar** [minad/vertico#425] [Ideas]
The context for this is that in my config I have,
```emacs-lisp
(cl-pushnew '(consult-hoogle buffer) vertico-multiform-categories)
```
However `vertico-buffer-mode` shows a modeline which I wanted to hide. So, I tried,
```emacs-lisp
(cl-pushnew '(consult-hoogle buffer (mode-line-format)) vertico-multiform-categories)
```
However this doesn't work since `cl-defmethod` for `vertico-buffer-mode` sets the `mode-line-format` during setup. The only way I could think of to hide modeline was to add another  method,
```emacs-lisp
(cl-defmethod vertico--setup :after
      (&context (vertico-buffer-mode (eql t))
                ((completion-metadata-get
                  (completion-metadata "" minibuffer-completion-table nil) 'category)
                 (eql 'consult-hoogle)))
      (setq-local mode-line-format nil
                  vertico-count (1+ vertico-count)))
```
which is cumbersome. 
Are there any variables which must be set before calling `vertico--setup`? If so to keep things simple the best way might be to set them both before and after the setup so that the setup doesn't overwrite user customization. I can submit a patch for this.

**Comment by minad** [comment]
I think it is implemented like this for simplicity. We enable the modes before setup and do the same with the variables. Do you want to hide the mode line always for buffers or only for hoogle? If you want to override this always a custom cl-defmethod may indeed be a good approach.

**Comment by aikrahguzar** [comment]
> Do you want to hide the mode line always for buffers or only for hoogle?

Only for hoogle but hoogle is also the only thing for which I am using `vertico-buffer-mode` since I want to display a buffer which shows the documentation for the current result on the side and I think it is not possible to split the mini-window vertically.

But I actually can't think of a way to write a custom `cl-defmethod` to always hide the modeline. I don't understand generics very well but I think the method I wrote works only because it is more specific than the one in `vertico-buffer`. But I wouldn't know how to ensure the ordering if the context was just `(vertico-buffer-mode (eql t))`.

**Reply by minad** [reply]
> I don't understand generics very well but I think the method I wrote works only because it is more specific than the one in vertico-buffer. 

That's right. There is also the `:extra` keyword which you can use. I think the following should always work:

~~~ elisp
(with-eval-after-load 'vertico-buffer
  (cl-defmethod vertico--setup
    :extra "hide mode line"
    :after (&context (vertico-buffer-mode (eql t)))
    (setq-local mode-line-format nil)))
~~~

**Comment by aikrahguzar** [comment]
> ```emacs-lisp
> (with-eval-after-load 'vertico-buffer
>   (cl-defmethod vertico--setup
>     :extra "hide mode line"
>     :after (&context (vertico-buffer-mode (eql t)))
>     (setq-local mode-line-format nil)))
> ```

Thanks a lot! This works. BTW, I can't figure out from the doc string of `cl-defmethod` what is the order of `:extra` method calls? Should they be considered more specific than the method with same qualifiers and specializers but without the `:extra` keyword or are the supposed to commute with them?

**Reply by minad** [reply]
I don't know. Probably the order of definition matters.

**Comment by aikrahguzar** [comment]
I still think that setting the variables both before and after `vertico--setup` will not add much complexity and will provide better ux especially since `vertico--setup` is a private function and `vertico-multiform-mode` is the official way of configuring per category and per command options.

**Reply by minad** [reply]
> I still think that setting the variables both before and after vertico--setup will not add much complexity...

This issue hasn't come up often, so I am not enthusiastic about adding a special handling for variables. Right now everything is executed in order before setup, modes, variables, etc. I've also considered to add a vertico-buffer-mode-line customization option, but this also probably not worth it.

**Comment by aikrahguzar** [comment]
> This issue hasn't come up often, so I am not enthusiastic about adding a special handling for variables. Right now everything is executed in order before setup, modes, variables, etc. I've also considered to add a vertico-buffer-mode-line customization option, but this also probably not worth it.

Makes sense! Thanks again!
